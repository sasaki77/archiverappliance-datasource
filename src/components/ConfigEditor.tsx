import React, { PureComponent, ChangeEvent } from 'react';
import { DataSourceHttpSettings, InlineSwitch, InlineField, Combobox, ComboboxOption } from '@grafana/ui';
import { DataSourcePluginOptionsEditorProps } from '@grafana/data';
import { AADataSourceOptions, operatorList } from '../types';
import { toComboboxOption } from './utils';

const LABEL_WIDTH = 26;
const operatorOptions: Array<ComboboxOption<string>> = operatorList.map(toComboboxOption);

export type Props = DataSourcePluginOptionsEditorProps<AADataSourceOptions>;

export class ConfigEditor extends PureComponent<Props> {
  constructor(props: Props) {
    super(props);
  }

  onUseBEChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      useBackend: !options.jsonData.useBackend,
    };
    onOptionsChange({ ...options, jsonData });
  };

  onOperatorChange = (option: ComboboxOption) => {
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      defaultOperator: option.value,
    };
    onOptionsChange({ ...options, jsonData });
  };

  onHideInvalidChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      hideInvalid: !options.jsonData.hideInvalid,
    };
    onOptionsChange({ ...options, jsonData });
  };

  onUseLiveUpdateChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      useLiveUpdate: !options.jsonData.useLiveUpdate,
    };
    onOptionsChange({ ...options, jsonData });
  };

  onLiveUpdateURIChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      liveUpdateURI: event.target.value,
    };
    onOptionsChange({ ...options, jsonData });
  };

  render() {
    const { options, onOptionsChange } = this.props;

    return (
      <>
        <DataSourceHttpSettings
          defaultUrl="http://localhost:17668/retrieval"
          dataSourceConfig={options}
          onChange={onOptionsChange}
        />
        <h3 className="page-heading">Misc</h3>
        <InlineField
          label="Use Backend"
          labelWidth={LABEL_WIDTH}
          tooltip="Checking this option will enable the data retrieval with backend. The archived data is retrieved and processed on Grafana server, then the data is sent to Grafana client."
          interactive={true}
        >
          <InlineSwitch value={options.jsonData.useBackend ?? false} onChange={this.onUseBEChange} />
        </InlineField>
        <InlineField
          label="Default Operator"
          labelWidth={LABEL_WIDTH}
          interactive={true}
          tooltip={
            <p>
              Controls processing of data during data retrieval. Refer{' '}
              <a
                href="https://epicsarchiver.readthedocs.io/en/latest/user/userguide.html#processing-of-data"
                target="_blank"
                rel="noopener noreferrer"
              >
                Archiver Appliance User Guide
              </a>{' '}
              about processing of data. Special operator <code>raw</code> and <code>last</code> are also available.{' '}
              <code>raw</code> allows to retrieve the data without processing. <code>last</code> allows to retrieve the
              last data in the specified time range.
            </p>
          }
        >
          <Combobox
            value={options.jsonData.defaultOperator}
            options={operatorOptions}
            onChange={this.onOperatorChange}
            placeholder="mean"
          />
        </InlineField>
        <InlineField
          label="Hide Invalid"
          labelWidth={LABEL_WIDTH}
          interactive={true}
          tooltip={<p>Hide sample data whose severity is invalid with a null value.</p>}
        >
          <InlineSwitch value={options.jsonData.hideInvalid ?? false} onChange={this.onHideInvalidChange} />
        </InlineField>
        <InlineField
          label="Use live feature (Alpha)"
          labelWidth={LABEL_WIDTH}
          interactive={true}
          tooltip="(Caution) This is a alpha feature. Live feature provides live updating with PVWS WebSocket server."
        >
          <InlineSwitch value={options.jsonData.useLiveUpdate ?? false} onChange={this.onUseLiveUpdateChange} />
        </InlineField>
        <InlineField
          label="PVWS URI (Alpha)"
          labelWidth={LABEL_WIDTH}
          interactive={true}
          tooltip="URI for PVWS WebSocket server."
        >
          <input
            type="text"
            value={options.jsonData.liveUpdateURI}
            className="gf-form-input max-width-15"
            placeholder="ws://localhost:8080/pvws/pv"
            onChange={this.onLiveUpdateURIChange}
          />
        </InlineField>
      </>
    );
  }
}
