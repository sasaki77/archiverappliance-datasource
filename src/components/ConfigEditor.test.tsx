import React from 'react';
import { render } from '@testing-library/react';

import { ConfigEditor, Props } from './ConfigEditor';

const setup = (propOverrides?: object) => {
  const props: Props = {
    options: {
      id: 1,
      orgId: 1,
      uid: '',
      typeLogoUrl: '',
      name: 'ArchiverAppliance',
      access: 'proxy',
      url: '',
      database: '',
      type: 'ArchiverAppliance',
      typeName: 'ArchiverAppliance',
      user: '',
      basicAuth: false,
      basicAuthUser: '',
      isDefault: true,
      readOnly: false,
      withCredentials: false,
      secureJsonFields: {
        accessKey: false,
        secretKey: false,
      },
      jsonData: {},
      secureJsonData: {
        secretKey: '',
        accessKey: '',
      },
    },
    onOptionsChange: jest.fn(),
  };

  Object.assign(props, propOverrides);

  return render(<ConfigEditor {...props} />);
};

describe('Render', () => {
  it('should render component', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
