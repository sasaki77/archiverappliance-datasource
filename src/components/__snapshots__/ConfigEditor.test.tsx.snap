// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="gf-form-group"
      >
        <h3
          class="page-heading"
        >
          HTTP
        </h3>
        <div
          class="gf-form-group"
        >
          <div
            class="gf-form"
          >
            <div
              class="css-aq4c43"
            >
              <label
                class="gf-form-label width-13"
              >
                URL
                <div
                  class="css-wf08df-Icon"
                  tabindex="0"
                >
                  <svg
                    data-testid="info-circle"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
              </label>
              <div
                style="flex-grow: 1;"
              >
                <input
                  aria-label="Datasource HTTP settings url"
                  class="gf-form-input width-20 css-y1sxu8"
                  placeholder="http://localhost:17668/retrieval"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="gf-form-group"
          >
            <div
              class="gf-form"
            >
              <label
                class="gf-form-label width-13"
              >
                Allowed cookies
                <div
                  class="css-wf08df-Icon"
                  tabindex="0"
                >
                  <svg
                    data-testid="info-circle"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
              </label>
              <div
                class="css-1k4ludt"
              >
                <div
                  class="css-xcstkt-input-wrapper"
                  data-testid="input-wrapper"
                >
                  <div
                    class="css-1w5c5dq-input-inputWrapper"
                  >
                    <input
                      class="css-1mlczho-input-input"
                      placeholder="New tag (enter key to add)"
                      style="padding-right: 12px;"
                      value=""
                    />
                    <div
                      class="css-7y3u6k-input-suffix"
                    >
                      <button
                        class="css-7e4tcf-button"
                        disabled=""
                        type="button"
                      >
                        <span
                          class="css-1mhnkuh"
                        >
                          Add
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="gf-form"
            >
              <div
                class="css-aq4c43"
              >
                <label
                  class="gf-form-label width-13"
                >
                  Timeout
                  <div
                    class="css-wf08df-Icon"
                    tabindex="0"
                  >
                    <svg
                      data-testid="info-circle"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </div>
                </label>
                <input
                  aria-label="Timeout in seconds"
                  class="gf-form-input width-20"
                  placeholder="Timeout in seconds"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <h3
          class="page-heading"
        >
          Auth
        </h3>
        <div
          class="gf-form-group"
        >
          <div
            class="gf-form-inline"
          >
            <div
              class="css-1f5gueu"
            >
              <label
                class="css-1mfv0aj"
                for="http-settings-basic-auth"
              >
                Basic auth
              </label>
              <div
                class="css-1rfunm5"
              >
                <div
                  class="css-15rfc12"
                >
                  <div
                    class="css-1so7yeh"
                  >
                    <input
                      id="http-settings-basic-auth"
                      type="checkbox"
                    />
                    <label
                      aria-label="Toggle switch"
                      for="http-settings-basic-auth"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-1f5gueu"
            >
              <label
                class="css-1mfv0aj"
                for="http-settings-with-credentials"
              >
                With Credentials
                <div
                  class="css-wf08df-Icon"
                  tabindex="0"
                >
                  <svg
                    data-testid="info-circle"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
              </label>
              <div
                class="css-1rfunm5"
              >
                <div
                  class="css-15rfc12"
                >
                  <div
                    class="css-1so7yeh"
                  >
                    <input
                      id="http-settings-with-credentials"
                      type="checkbox"
                    />
                    <label
                      aria-label="Toggle switch"
                      for="http-settings-with-credentials"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="gf-form-inline"
          >
            <div
              class="css-1f5gueu"
            >
              <label
                class="css-1mfv0aj"
                for="http-settings-tls-client-auth"
              >
                TLS Client Auth
              </label>
              <div
                class="css-1rfunm5"
              >
                <div
                  class="css-15rfc12"
                >
                  <div
                    class="css-1so7yeh"
                  >
                    <input
                      id="http-settings-tls-client-auth"
                      type="checkbox"
                    />
                    <label
                      aria-label="Toggle switch"
                      for="http-settings-tls-client-auth"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-1f5gueu"
            >
              <label
                class="css-1mfv0aj"
                for="http-settings-ca-cert"
              >
                With CA Cert
                <div
                  class="css-wf08df-Icon"
                  tabindex="0"
                >
                  <svg
                    data-testid="info-circle"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
              </label>
              <div
                class="css-1rfunm5"
              >
                <div
                  class="css-15rfc12"
                >
                  <div
                    class="css-1so7yeh"
                  >
                    <input
                      id="http-settings-ca-cert"
                      type="checkbox"
                    />
                    <label
                      aria-label="Toggle switch"
                      for="http-settings-ca-cert"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="gf-form-inline"
          >
            <div
              class="css-1f5gueu"
            >
              <label
                class="css-1mfv0aj"
                for="http-settings-skip-tls-verify"
              >
                Skip TLS Verify
              </label>
              <div
                class="css-1rfunm5"
              >
                <div
                  class="css-15rfc12"
                >
                  <div
                    class="css-1so7yeh"
                  >
                    <input
                      id="http-settings-skip-tls-verify"
                      type="checkbox"
                    />
                    <label
                      aria-label="Toggle switch"
                      for="http-settings-skip-tls-verify"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="gf-form-inline"
          >
            <div
              class="css-1f5gueu"
            >
              <label
                class="css-1mfv0aj"
                for="http-settings-forward-oauth"
              >
                Forward OAuth Identity
                <div
                  class="css-wf08df-Icon"
                  tabindex="0"
                >
                  <svg
                    data-testid="info-circle"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
              </label>
              <div
                class="css-1rfunm5"
              >
                <div
                  class="css-15rfc12"
                >
                  <div
                    class="css-1so7yeh"
                  >
                    <input
                      id="http-settings-forward-oauth"
                      type="checkbox"
                    />
                    <label
                      aria-label="Toggle switch"
                      for="http-settings-forward-oauth"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="gf-form-group"
        >
          <div
            class="gf-form"
          >
            <h6>
              Custom HTTP Headers
            </h6>
          </div>
          <div />
          <div
            class="gf-form"
          >
            <button
              class="css-b2ba3d-button"
              type="button"
            >
              <div
                class="css-wf08df-Icon"
              >
                <svg
                  data-testid="plus"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
              <span
                class="css-1mhnkuh"
              >
                Add header
              </span>
            </button>
          </div>
        </div>
      </div>
      <h3
        class="page-heading"
      >
        Misc
      </h3>
      <div
        class="gf-form-group"
      >
        <div
          class="gf-form-inline"
        >
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
            >
              Use Backend
              <div
                class="css-wf08df-Icon"
                tabindex="0"
              >
                <svg
                  data-testid="info-circle"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15rfc12"
              >
                <div
                  class="css-1so7yeh"
                >
                  <input
                    id="switch-1"
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle switch"
                    for="switch-1"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="gf-form-inline"
        >
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
            >
              Default Operator
              <div
                class="css-wf08df-Icon"
                tabindex="0"
              >
                <svg
                  data-testid="info-circle"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15fuo2f-input-wrapper css-1kstem6-SelectContainer"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-2-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="css-1scgfe8"
                >
                  <div
                    class="css-1kl463j-grafana-select-value-container"
                  >
                    <div
                      class=" css-1yc0yww-placeholder"
                      id="react-select-2-placeholder"
                    >
                      mean
                    </div>
                    <div
                      class=" css-8r78wt-Input"
                      data-value=""
                    >
                      <input
                        aria-autocomplete="list"
                        aria-describedby="react-select-2-placeholder"
                        aria-expanded="false"
                        aria-haspopup="true"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class=""
                        id="react-select-2-input"
                        role="combobox"
                        spellcheck="false"
                        style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="css-uvldi-input-suffix"
                  >
                    <div
                      class="css-wf08df-Icon"
                    >
                      <svg
                        data-testid="angle-down"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="gf-form-group"
    >
      <h3
        class="page-heading"
      >
        HTTP
      </h3>
      <div
        class="gf-form-group"
      >
        <div
          class="gf-form"
        >
          <div
            class="css-aq4c43"
          >
            <label
              class="gf-form-label width-13"
            >
              URL
              <div
                class="css-wf08df-Icon"
                tabindex="0"
              >
                <svg
                  data-testid="info-circle"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </label>
            <div
              style="flex-grow: 1;"
            >
              <input
                aria-label="Datasource HTTP settings url"
                class="gf-form-input width-20 css-y1sxu8"
                placeholder="http://localhost:17668/retrieval"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="gf-form-group"
        >
          <div
            class="gf-form"
          >
            <label
              class="gf-form-label width-13"
            >
              Allowed cookies
              <div
                class="css-wf08df-Icon"
                tabindex="0"
              >
                <svg
                  data-testid="info-circle"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </label>
            <div
              class="css-1k4ludt"
            >
              <div
                class="css-xcstkt-input-wrapper"
                data-testid="input-wrapper"
              >
                <div
                  class="css-1w5c5dq-input-inputWrapper"
                >
                  <input
                    class="css-1mlczho-input-input"
                    placeholder="New tag (enter key to add)"
                    style="padding-right: 12px;"
                    value=""
                  />
                  <div
                    class="css-7y3u6k-input-suffix"
                  >
                    <button
                      class="css-7e4tcf-button"
                      disabled=""
                      type="button"
                    >
                      <span
                        class="css-1mhnkuh"
                      >
                        Add
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="gf-form"
          >
            <div
              class="css-aq4c43"
            >
              <label
                class="gf-form-label width-13"
              >
                Timeout
                <div
                  class="css-wf08df-Icon"
                  tabindex="0"
                >
                  <svg
                    data-testid="info-circle"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
              </label>
              <input
                aria-label="Timeout in seconds"
                class="gf-form-input width-20"
                placeholder="Timeout in seconds"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <h3
        class="page-heading"
      >
        Auth
      </h3>
      <div
        class="gf-form-group"
      >
        <div
          class="gf-form-inline"
        >
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
              for="http-settings-basic-auth"
            >
              Basic auth
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15rfc12"
              >
                <div
                  class="css-1so7yeh"
                >
                  <input
                    id="http-settings-basic-auth"
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle switch"
                    for="http-settings-basic-auth"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
              for="http-settings-with-credentials"
            >
              With Credentials
              <div
                class="css-wf08df-Icon"
                tabindex="0"
              >
                <svg
                  data-testid="info-circle"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15rfc12"
              >
                <div
                  class="css-1so7yeh"
                >
                  <input
                    id="http-settings-with-credentials"
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle switch"
                    for="http-settings-with-credentials"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="gf-form-inline"
        >
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
              for="http-settings-tls-client-auth"
            >
              TLS Client Auth
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15rfc12"
              >
                <div
                  class="css-1so7yeh"
                >
                  <input
                    id="http-settings-tls-client-auth"
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle switch"
                    for="http-settings-tls-client-auth"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
              for="http-settings-ca-cert"
            >
              With CA Cert
              <div
                class="css-wf08df-Icon"
                tabindex="0"
              >
                <svg
                  data-testid="info-circle"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15rfc12"
              >
                <div
                  class="css-1so7yeh"
                >
                  <input
                    id="http-settings-ca-cert"
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle switch"
                    for="http-settings-ca-cert"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="gf-form-inline"
        >
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
              for="http-settings-skip-tls-verify"
            >
              Skip TLS Verify
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15rfc12"
              >
                <div
                  class="css-1so7yeh"
                >
                  <input
                    id="http-settings-skip-tls-verify"
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle switch"
                    for="http-settings-skip-tls-verify"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="gf-form-inline"
        >
          <div
            class="css-1f5gueu"
          >
            <label
              class="css-1mfv0aj"
              for="http-settings-forward-oauth"
            >
              Forward OAuth Identity
              <div
                class="css-wf08df-Icon"
                tabindex="0"
              >
                <svg
                  data-testid="info-circle"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </div>
            </label>
            <div
              class="css-1rfunm5"
            >
              <div
                class="css-15rfc12"
              >
                <div
                  class="css-1so7yeh"
                >
                  <input
                    id="http-settings-forward-oauth"
                    type="checkbox"
                  />
                  <label
                    aria-label="Toggle switch"
                    for="http-settings-forward-oauth"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="gf-form-group"
      >
        <div
          class="gf-form"
        >
          <h6>
            Custom HTTP Headers
          </h6>
        </div>
        <div />
        <div
          class="gf-form"
        >
          <button
            class="css-b2ba3d-button"
            type="button"
          >
            <div
              class="css-wf08df-Icon"
            >
              <svg
                data-testid="plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
            <span
              class="css-1mhnkuh"
            >
              Add header
            </span>
          </button>
        </div>
      </div>
    </div>
    <h3
      class="page-heading"
    >
      Misc
    </h3>
    <div
      class="gf-form-group"
    >
      <div
        class="gf-form-inline"
      >
        <div
          class="css-1f5gueu"
        >
          <label
            class="css-1mfv0aj"
          >
            Use Backend
            <div
              class="css-wf08df-Icon"
              tabindex="0"
            >
              <svg
                data-testid="info-circle"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </label>
          <div
            class="css-1rfunm5"
          >
            <div
              class="css-15rfc12"
            >
              <div
                class="css-1so7yeh"
              >
                <input
                  id="switch-1"
                  type="checkbox"
                />
                <label
                  aria-label="Toggle switch"
                  for="switch-1"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="gf-form-inline"
      >
        <div
          class="css-1f5gueu"
        >
          <label
            class="css-1mfv0aj"
          >
            Default Operator
            <div
              class="css-wf08df-Icon"
              tabindex="0"
            >
              <svg
                data-testid="info-circle"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              />
            </div>
          </label>
          <div
            class="css-1rfunm5"
          >
            <div
              class="css-15fuo2f-input-wrapper css-1kstem6-SelectContainer"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class="css-1scgfe8"
              >
                <div
                  class="css-1kl463j-grafana-select-value-container"
                >
                  <div
                    class=" css-1yc0yww-placeholder"
                    id="react-select-2-placeholder"
                  >
                    mean
                  </div>
                  <div
                    class=" css-8r78wt-Input"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-2-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="css-uvldi-input-suffix"
                >
                  <div
                    class="css-wf08df-Icon"
                  >
                    <svg
                      data-testid="angle-down"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
